api-mock/fetcher.ts

```
@@ -0,0 +1,248 @@
+                headers: responseHeaders,
+                text,
+            });
+        } catch (error) {
+            callback?.(error, null);
+        }
+    }
+
+    function post(pathValue: string): ChainedRequest {
+        path = pathValue;
+        method = "post";
+        return this;
+    }
+
+    function get(pathValue: string): ChainedRequest {
+        path = pathValue;
+        method = "get";
+        return this;
+    }
+
+    function set(header: string, value: string): ChainedRequest {
+        headers[header] = value;
+        return this;
+    }
+
+    function send(data: any): ChainedRequest {
+        input = data;
+        return this;
+    }
+
+    function expect(status: number, cb?: Callback): ChainedRequest {
+        expectedStatus = status;
+        if (cb !== undefined) {
+            this.end(cb);
+        }
+        return this;
+    }
+
+    function end(cb: Callback) {
+        callback = cb;
+        executeRequest();
+    }
+
+    return {
+        post,
+        get,
+        set,
+        send,
+        expect,
+        end,
+    };
+}
+
+export async function queryAPI({
+    method,
+    path,
+    input,
+    headers,
+    returnResponse,
+    skipInit,
+}: {
+    method: "post" | "get" | "delete" | "put";
+    path: string;
+    input?: object;
+    headers?: Record<string, string>;
+    returnResponse?: boolean;
+    skipInit?: boolean;
+}) {
+    if (!skipInit && status === "NOT_READY") {
+        await initApp();
+    }
+    try {
+        let response = await fetch(`http://localhost:${apiMockPort}${path}`, {
+            method,
+            headers: {
+                "Content-Type": "application/json",
+                ...headers,
+            },
+            body: JSON.stringify(input),
+        });
+
+        if (returnResponse) {
+            return response;
+        }
+
+        if (!response.ok) {
+            throw response;
+        }
+
+        return await response.json().catch(() => undefined);
```

Are there cases where we expect non-json responses? If not, we should just let this throw.

---

@sandrodesouza I think this is the only remaining comment on this.

---

yes, there is this case here EmailVerificationRecipe.getInstance().updateSessionIfRequiredPostEmailVerification

---

that also seems to return through `res.json`

---

yes but could be undefined

---

I see, I thought you wrapped those in objects, but I guess this works as well.